spec/models/package_spec.rb

```diff
@@ -169,20 +169,269 @@
   describe 'dispatch_stockit_item' do
     let(:package) { create :package, :with_set_item }
     let!(:location) { create :location, :dispatched }
+    let!(:packages_location) { create :packages_location, location: location, package: package}
     before { expect(Stockit::ItemSync).to receive(:dispatch).with(package) }
 
     it 'set dispatch related details' do
       package.dispatch_stockit_item
-      expect(package.location).to eq(location)
+      expect(package.locations.first).to eq(location)
       expect(package.stockit_sent_on).to_not be_nil
     end
 
     it 'update set relation on dispatching single package' do
-      sibling_package = create :package, :with_set_item, item: package.item
+      sibling_package = create :package, :with_set_item, :package_with_locations, item: package.item
       package.dispatch_stockit_item
       package.save
       expect(package.set_item_id).to be_nil
       expect(sibling_package.reload.set_item_id).to be_nil
     end
   end
+
+  describe '#add_location' do
+    it 'adds location to package if not added' do
+      location = create :location
+      package  = create :package
+      package.add_location(location.id)
+      expect(package.locations).to include(location)
+    end
+
+    it 'do not add location to package if same location already exist' do
+      package = create :package, :package_with_locations
+      location = package.packages_locations.first.location
+      expect(package.add_location(location.id)).to be_nil
+      expect(package.locations).to include(location)
+    end
+  end
+
+  describe '#update_or_create_qty_moved_to_location' do
+    let!(:package) { create :package }
+    let!(:location) { create :location }
+
+    it 'creates associated packages_location record if we do not have packages_location record with provided location_id' do
+      expect{
+        package.update_or_create_qty_moved_to_location(location.id, 10)
```

Comment:
Will this fail if package doesn't have at least quantity 10? Perhaps it could be `package.quantity` or similar instead?